src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp

```diff
@@ -1132,6 +1132,7 @@ MergeTreeData::MutableDataPartPtr MergeTreeDataMergerMutator::mutatePartToTempor
         future_part.name, future_part.type, future_part.part_info, single_disk_volume, "tmp_mut_" + future_part.name);
 
     new_data_part->is_temp = true;
+    new_data_part->uuid = source_part->uuid;
```

Comment:
mmm, but mutation produce a different part? Why use the same UUID?

Reply:
That is a very good question, maybe even a philosophical one. Although mutations produces new "physical" we want still to consider it to be the same "logical" part. To be more precise in the definition the UUID doesn't actually identify a part, but instead identifies the initial data from which the part was formed when it was initially created.

Let's say we "lock" part uuid 1 on a server.

A. If we choose to generate new UUIDs on all mutations then we need to block/reject (or skip things like TTLs) operations, including ALTERs with DELETEs, TTLs for recompression, moving parts between disks, otherwise new UUIDs would break de-duplication logic.

B. If we choose not to generate new UUIDs on mutations then we can allow ALTER UPDATE/DELETEs, TTLs, ... to proceed as normal. The only requirement is that the part structure shouldn't change, ie it should still be "ATTACHABLE" on the destination.

B has one problem (which can be a big deal in some cases), a logical race condition. If we allow mutations to run then we need to "freeze" the original part before movement so it doesn't disappear suddenly. Then running an ALTER DELETE on the source would appear to have deleted some data, but when the original part is attached on the destination server the data would suddenly "re-appear".

Now after giving it some more thought A seems more sensible: Reject all user initiated ALTERs, skip background operations if a part is "locked for movement".

Reply:
+1 for A for simplicity